python/taichi/lang/kernel_arguments.py

```diff
@@ -36,6 +36,29 @@ def extract(self, x):
 """
 
 
+class ArgAnyArray:
+    """Type annotation for arbitrary arrays, including external arrays and Taichi ndarrays."""
+    def extract(self, x):
+        return to_taichi_type(x.dtype), len(x.shape)
```

Comment:
TODO: We should include information like AOS/SOA, matrix shape

Reply:
BTW should I support these in `ext_arr()` first or just implement them in `any_arr()` and deprecate `ext_arr()` in the future?

Reply:
>  just implement them in any_arr()

This sounds good